### PR TITLE
test(consolidation): direct greedy_cluster coverage + scope_id vote tests (#440, #441, #442)

### LINKED ISSUE #440: [super-qa] medium: greedy_cluster() has no direct unit tests; edge cases not exercised
> ⚠️ Generated by `/super-qa` 2026-06-01. Severity & details from agent analysis; **not independently source-verified**. Review before action.

**greedy_cluster() has no direct unit tests; edge cases not exercised**

- **ID:** `consolidation/testing-greedy-cluster-no-direct-tests`
- **Severity:** medium
- **Area:** consolidation · **Type:** test
- **Location:** `src/consolidation/cluster.rs:106-140`
- **Auto-fixable:** no

## Description
The `greedy_cluster()` function is private and only tested indirectly through `cluster_fusion()`. Direct unit tests are absent for: (1) empty input (`facts = []` → should return `vec![]`); (2) single-element input (→ one cluster of size 1); (3) threshold = 1.0 exactly (all facts in separate singletons unless embedding identical); (4) transitivity: A∼B, B∼C but A≁C — does C end up in A's cluster via B? The current greedy expansion would include C (since it expands from each member), but this is not asserted. All four cluster tests use only `CLUSTER_SIMILARITY_THRESHOLD` (0.85) with hand-crafted embeddings at dim=4.

## Suggested approach
Add a `#[cfg(test)] mod greedy_cluster_tests` in cluster.rs that calls `greedy_cluster()` directly (it is private so tests are in the same file). Cover: empty slice, single element, threshold=0.0 (all in one cluster), threshold=1.0 (all singletons unless identical), and the transitive chaining case (A≃B, B≃C, A≁C).

## References
—

### LINKED ISSUE #441: [super-qa] medium: Majority-vote scope_id selection (including tie-break) has no test
> ⚠️ Generated by `/super-qa` 2026-06-01. Severity & details from agent analysis; **not independently source-verified**. Review before action.

**Majority-vote scope_id selection (including tie-break) has no test**

- **ID:** `consolidation/testing-scope-majority-vote-untested`
- **Severity:** medium
- **Area:** consolidation · **Type:** test
- **Location:** `src/consolidation/cluster.rs:73-85`
- **Auto-fixable:** no

## Description
The cluster scope_id is computed by a majority vote across source facts (cluster.rs:76-84), with the deterministic tie-break: 'equal counts → lowest scope_id wins' (implemented via `.then_with(|| b.0.cmp(&a.0))`). All existing cluster tests (`cluster_formation`, `min_cluster_size_respected`, `cluster_summaries_stored`, `idempotent_rebuild`) insert facts with the hardcoded `scope_id: 1`. The comment in `tests/eval/quality/consolidation.rs:82` says 'Use scope to verify majority-vote scope propagation' but all 8 facts in that test share the same scope (`project:rust`). Neither the majority path nor the tie-break is ever verified to produce the correct `scope_id` on the resulting summary.

## Suggested approach
Add a unit test in cluster.rs that inserts facts split across two scopes (e.g., 3 facts in scope 10 and 2 facts in scope 20), runs `cluster_fusion`, retrieves the resulting summary via `SummaryStore::list_by_level`, and asserts `summary.scope_id == 10`. Add a second test with an even split (2 and 2) across scope IDs 10 and 5, and assert `summary.scope_id == 5` (lowest wins).

## References
—

### LINKED ISSUE #442: [super-qa] medium: greedy_cluster() has algebraic invariants well-suited for property-based testing
> ⚠️ Generated by `/super-qa` 2026-06-01. Severity & details from agent analysis; **not independently source-verified**. Review before action.

**greedy_cluster() has algebraic invariants well-suited for property-based testing**

- **ID:** `consolidation/testing-proptest-greedy-cluster-invariants`
- **Severity:** medium
- **Area:** consolidation · **Type:** test
- **Location:** `src/consolidation/cluster.rs:106-140`
- **Auto-fixable:** no

## Description
`proptest` is already a dev-dependency (Cargo.toml:41) and is used in `src/search/hybrid.rs` and `src/search/vector.rs`. `greedy_cluster()` has strong structural invariants that example-based tests cannot adequately cover: (1) **partition**: every index 0..n appears in exactly one cluster; (2) **within-cluster similarity**: for any two facts in the same cluster, there exists a chain of pairwise-similar facts connecting them; (3) **between-cluster dissimilarity**: no fact in cluster A is similar (> threshold) to a fact in cluster B; (4) **monotonicity**: lowering the threshold never reduces cluster count. None of these are verified.

## Suggested approach
Add a `#[cfg(test)] mod proptest_greedy_cluster` block in cluster.rs (mirroring `proptest_rrf` in hybrid.rs). Generate random `Vec<Fact>` with random unit-normalized embeddings and random thresholds in [0.0, 1.0]. Assert partition invariant: `clusters.iter().flatten().sorted().collect::<Vec<_>>() == (0..n).collect::<Vec<_>>()`. Assert no index appears twice. The similarity-chain and monotonicity properties can be checked with bounded exhaustive loops over the generated data.

## References
—

### PR DESCRIPTION
**Pull request by dutiona**
## What

Wave 6 of the **#253 consolidation sweep** — cluster-pass test coverage (test-only, no production change).

- **#440** — direct `greedy_cluster` unit tests: empty, single, threshold 0.0 (collapse), threshold 1.0 (strict `>` ⇒ all singletons incl. identical), transitive single-linkage chain (A~B, B~C, A≁C ⇒ merged via B).
- **#441** — majority-vote `scope_id`: 3-vs-2 → majority; 2-vs-2 tie → lowest scope_id (deterministic tie-break). Fixtures create the FK scope rows explicitly.
- **#442** — proptest partition invariant: random embeddings + any threshold ⇒ every index in exactly one cluster.

## Verification

`cargo test --workspace --all-features` 37 binaries pass (13 cluster tests, up from 5); `clippy -D warnings` + `fmt` + `cargo deny` clean.

Closes #440. Closes #441. Closes #442. Part of #253.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/consolidation/cluster.rs [hunk 1]
```diff
@@ -354,4 +354,215 @@ mod tests {
...
+            scope_id,
+            importance: 0.5,
+            access_count: 0,
+            last_accessed: Utc::now(),
+            metadata: serde_json::json!({}),
+            is_pinned: false,
+            importance_score: Fact::UNSCORED_IMPORTANCE,
+            surfaced_at: None,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (dutiona) [reply src/consolidation/cluster.rs:383]: Declining: `Utc::now()` per-invocation in these test helpers mirrors the established convention of every sibling helper in this module (`insert_fact`, the `mod.rs`/`dedup.rs` fixtures) and the rest of the crate's tests. These tests never assert on timestamps, so 'temporal consistency' is irrelevant, and a handful of clock calls in a unit test is not a meaningful cost. Pre-threading a timestamp would make these helpers inconsistent with their neighbors for no behavioral or perf benefit. Resolving.

File: src/consolidation/cluster.rs [hunk 2]
```diff
@@ -354,4 +354,215 @@ mod tests {
...
+                source_event_id: None,
+                scope_id,
+                importance: 0.5,
+                access_count: 0,
+                last_accessed: Utc::now(),
+                metadata: serde_json::json!({}),
+                is_pinned: false,
+            })
+            .unwrap()
+    }
```
[the remark this answers is not in the document]
Comment (dutiona) [reply src/consolidation/cluster.rs:468]: Declining: `Utc::now()` per-invocation in these test helpers mirrors the established convention of every sibling helper in this module (`insert_fact`, the `mod.rs`/`dedup.rs` fixtures) and the rest of the crate's tests. These tests never assert on timestamps, so 'temporal consistency' is irrelevant, and a handful of clock calls in a unit test is not a meaningful cost. Pre-threading a timestamp would make these helpers inconsistent with their neighbors for no behavioral or perf benefit. Resolving.

File: src/consolidation/cluster.rs [hunk 3]
```diff
@@ -354,4 +354,215 @@ mod tests {
...
+        create_scope(&conn, 20, "twenty");
+        // 3 facts in scope 10, 2 in scope 20 → majority scope 10 wins.
+        for i in 0..3 {
+            insert_fact_scoped(&conn, dim, &format!("a{i}"), near(i), 10);
+        }
+        for i in 3..5 {
+            insert_fact_scoped(&conn, dim, &format!("b{i}"), near(i), 20);
+        }
+        assert_eq!(single_cluster_scope(&conn, dim), 10);
+    }
```
[the remark this answers is not in the document]
Comment (dutiona) [reply src/consolidation/cluster.rs:520]: Declining: `Utc::now()` per-invocation in these test helpers mirrors the established convention of every sibling helper in this module (`insert_fact`, the `mod.rs`/`dedup.rs` fixtures) and the rest of the crate's tests. These tests never assert on timestamps, so 'temporal consistency' is irrelevant, and a handful of clock calls in a unit test is not a meaningful cost. Pre-threading a timestamp would make these helpers inconsistent with their neighbors for no behavioral or perf benefit. Resolving.

File: src/consolidation/cluster.rs [hunk 4]
```diff
@@ -354,4 +354,215 @@ mod tests {
...
+        create_scope(&conn, 5, "five");
+        // 2 facts in scope 10, 2 in scope 5 → tie → lowest scope_id (5) wins.
+        for i in 0..2 {
+            insert_fact_scoped(&conn, dim, &format!("a{i}"), near(i), 10);
+        }
+        for i in 2..4 {
+            insert_fact_scoped(&conn, dim, &format!("b{i}"), near(i), 5);
+        }
+        assert_eq!(single_cluster_scope(&conn, dim), 5);
+    }
```
[the remark this answers is not in the document]
Comment (dutiona) [reply src/consolidation/cluster.rs:537]: Declining: `Utc::now()` per-invocation in these test helpers mirrors the established convention of every sibling helper in this module (`insert_fact`, the `mod.rs`/`dedup.rs` fixtures) and the rest of the crate's tests. These tests never assert on timestamps, so 'temporal consistency' is irrelevant, and a handful of clock calls in a unit test is not a meaningful cost. Pre-threading a timestamp would make these helpers inconsistent with their neighbors for no behavioral or perf benefit. Resolving.
